File: fdlsandbox-android-java/app/src/main/java/com/omatt/fdlsandbox/modules/inappbrowser/InAppBrowserActivity.java
package com.omatt.fdlsandbox.modules.inappbrowser;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import com.omatt.fdlsandbox.R;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnCheckedChanged;

/**
 * Created by omarmatthew on 10/12/17.
 * Simulates an in-app browser
 */

public class InAppBrowserActivity extends AppCompatActivity {
    private final String TAG = "WebViewActivity";
    private String dynamicLink;
    @BindView(R.id.webview)
    WebView webView;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.layout_activity_in_app_browser);
        ButterKnife.bind(this);

        Bundle bundle = getIntent().getExtras();
        if(bundle != null) dynamicLink = bundle.getString("KEY_FDL");
        initWebView(false, dynamicLink);
    }

    // If true, the dynamic link will be forced to be load in the WebView
    @OnCheckedChanged(R.id.switch_override_webview)
    void onSwitchChanged(boolean checked) {
        Log.i(TAG, "switch " + checked);
        initWebView(checked, dynamicLink);
    }

    void initWebView(boolean overrideWebView, String dynamicLink) {
        if (overrideWebView) webView.setWebViewClient(new CustomWebViewClient());
        else webView.setWebViewClient(null);
        webView.loadData(
                "<html>\n" +
                        "<body>\n" +
                        "Click <a href=\"" + dynamicLink + "\">here</a> to go to your FDL.\n" +
                        "</body>\n" +
                        "</html>",
                "text/html; charset=utf-8", "UTF-8");
    }
}

// Override web view client implementation
class CustomWebViewClient extends WebViewClient {
    public boolean shouldOverrideUrlLoading(WebView view, String url) {
        // Load the url to the WebView
        view.loadUrl(url);
        // return true if this method handled the link event
        // or false otherwise
        return true;
    }
}

File: fdlsandbox-android-kotlin/app/src/main/java/com/omatt/fdlsandbox/modules/inappbrowser/InAppBrowserActivity.kt
package com.omatt.fdlsandbox.modules.inappbrowser

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import android.util.Log
import android.webkit.WebViewClient
import com.omatt.fdlsandbox.AppController
import com.omatt.fdlsandbox.R
import kotlinx.android.synthetic.main.layout_activity_in_app_browser.*
import javax.inject.Inject

/**
 * Created by omarmatthew on 10/14/2017.
 * Simulates FDL behavior on an in-app browser
 */
class InAppBrowserActivity : AppCompatActivity(), InAppBrowserContract.View {
    private val TAG = "InAppBrowserActivity"
    var checked = false
    lateinit var dynamicLink: String

    @Inject
    lateinit var inAppBrowserPresenter: InAppBrowserPresenter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.layout_activity_in_app_browser)
        AppController.component.inject(this)

        dynamicLink = intent.extras?.getString("KEY_FDL").toString()

        switch_override_webview.setOnCheckedChangeListener({ _, checked ->
            Log.i(TAG, "switch $checked")
            inAppBrowserPresenter.initWebView(checked, dynamicLink)
        })
    }

    override fun onResume() {
        super.onResume()
        inAppBrowserPresenter.takeView(this)
        inAppBrowserPresenter.initWebView(checked, dynamicLink)
    }

    // @OnCheckedChanged butterknife action doesn't seem to be triggered in Kotlin
    // If true, the dynamic link will be forced to be load in the WebView
//    @OnCheckedChanged(R.id.switch_override_webview)
//    internal fun onSwitchChanged(checked: Boolean) {
//        Log.i(TAG, "switch $checked")
//        inAppBrowserPresenter.initWebView(checked, dynamicLink)
//    }

    override fun loadUrl(url: String) {
        webview.loadUrl(url)
    }

    override fun setWebViewClient(webViewClient: WebViewClient?) {
        webview.webViewClient = webViewClient
    }

    override fun loadWebViewData(data: String, mimeType: String, encoding: String) {
        webview.loadData(data, mimeType, encoding)
    }
}